### PR TITLE
Support stat sliding window bucket count config of circuit breaker

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
### Describe what this PR does / why we need it
Support stat sliding window bucket count config of circuit breaker.

### Does this pull request fix one issue?
#334 

### Special notes for reviews

### CODE REVIEW

File: core/circuitbreaker/rule.go [hunk 1]
```diff
@@ -100,3 +108,16 @@ func (r *Rule) isEqualsTo(newRule *Rule) bool {
...
 }
+
+func (r *Rule) getValidStatSlidingWindowBucketCount() uint32 {
+	interval := r.StatIntervalMs
+	bucketCount := r.StatSlidingWindowBucketCount
+	if bucketCount == 0 {
+		bucketCount = 1
+	} else {
+		if interval%bucketCount != 0 {
+			bucketCount = 1
```
Comment (liqiangz) [inline core/circuitbreaker/rule.go:119]: It might be better to add a warning log here.

File: core/circuitbreaker/rule.go [hunk 2]
```diff
@@ -61,15 +68,16 @@ type Rule struct {
...
+		r.Id, r.Resource, r.Strategy, r.RetryTimeoutMs, r.MinRequestAmount, r.StatIntervalMs, r.StatSlidingWindowBucketCount, r.MaxAllowedRtMs, r.Threshold)
 }
 
 func (r *Rule) isStatReusable(newRule *Rule) bool {
 	if newRule == nil {
 		return false
 	}
-	return r.Resource == newRule.Resource && r.Strategy == newRule.Strategy && r.StatIntervalMs == newRule.StatIntervalMs
+	return r.Resource == newRule.Resource && r.Strategy == newRule.Strategy && r.StatIntervalMs == newRule.StatIntervalMs &&
+		r.StatSlidingWindowBucketCount == newRule.StatSlidingWindowBucketCount
```
Comment (liqiangz) [inline core/circuitbreaker/rule.go:80]: It should be modified in isEqualsToBase.

Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule.go:80]: Resolved. Thanks.

File: core/circuitbreaker/rule_manager.go [hunk 3]
```diff
@@ -374,5 +374,8 @@ func IsValid(r *Rule) error {
 	if r.Strategy == ErrorRatio && r.Threshold > 1.0 {
 		return errors.New("invalid error ratio threshold (valid range: [0.0, 1.0])")
 	}
+	if r.StatSlidingWindowBucketCount != 0 && r.StatIntervalMs%r.StatSlidingWindowBucketCount != 0 {
+		logging.Warn("The following must be true: StatIntervalMs % StatSlidingWindowBucketCount == 0. StatSlidingWindowBucketCount will be replaced by 1", "rule", r)
```
Comment (sanxun0325) [inline core/circuitbreaker/rule_manager.go:378]: Print the log according to the specification  https://github.com/alibaba/sentinel-golang/pull/293

Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule_manager.go]: Resolved. Thanks.

File: core/circuitbreaker/rule.go [hunk 4]
```diff
@@ -61,7 +61,14 @@ type Rule struct {
 	// that can trigger circuit breaking.
 	MinRequestAmount uint64 `json:"minRequestAmount"`
 	// StatIntervalMs represents statistic time interval of the internal circuit breaker (in ms).
+	// Currently the statistic is collected by sliding window.
```
Comment (louyuting) [inline core/circuitbreaker/rule.go:64]: `Currently the statistic interval is collected by sliding window.`

File: core/circuitbreaker/circuit_breaker.go [hunk 5]
```diff
@@ -237,8 +237,9 @@ func newSlowRtCircuitBreakerWithStat(r *Rule, stat *slowRequestLeapArray) *slowR
 
 func newSlowRtCircuitBreaker(r *Rule) (*slowRtCircuitBreaker, error) {
 	interval := r.StatIntervalMs
+	bucketCount := r.getValidStatSlidingWindowBucketCount()
```
Comment (louyuting) [inline core/circuitbreaker/circuit_breaker.go:240]: Use an independent func to calculate bucketCount.
Member function of Rule is not good

File: core/circuitbreaker/rule.go [hunk 6]
```diff
@@ -61,7 +61,14 @@ type Rule struct {
...
 	StatIntervalMs uint32 `json:"statIntervalMs"`
+	// StatSlidingWindowBucketCount represents the bucket count of statistic sliding window.
+	// The statistic will be more precise as the bucket count increases, but the memory cost increases too.
+	// The following must be true — “StatIntervalMs % StatSlidingWindowBucketCount == 0”,
+	// otherwise StatSlidingWindowBucketCount will be replaced by 1.
+	// If it is not set, default value 1 will be used.
+	StatSlidingWindowBucketCount uint32 `json:"slidingWindowBucketCount"`
```
Comment (louyuting) [inline core/circuitbreaker/rule.go:71]: statSlidingWindowBucketCount

File: core/circuitbreaker/rule_manager.go [hunk 7]
```diff
@@ -393,5 +393,8 @@ func IsValid(r *Rule) error {
 	if r.Strategy == ErrorRatio && r.Threshold > 1.0 {
 		return errors.New("invalid error ratio threshold (valid range: [0.0, 1.0])")
 	}
+	if r.StatSlidingWindowBucketCount != 0 && r.StatIntervalMs%r.StatSlidingWindowBucketCount != 0 {
```
Comment (louyuting) [inline core/circuitbreaker/rule_manager.go:396]: How about `r.StatSlidingWindowBucketCount != 0` but `r.StatIntervalMs==0`?

Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule_manager.go:396]: `StatIntervalMs` already checked previously.

Comment (louyuting) [reply core/circuitbreaker/rule_manager.go:396]: okay.

File: core/circuitbreaker/circuit_breaker.go (hunk 5 above)
Comment (luckyxiaoqiang) [reply core/circuitbreaker/circuit_breaker.go]: Updated.

File: core/circuitbreaker/rule.go (hunk 4 above)
Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule.go]: Updated.

File: core/circuitbreaker/rule.go (hunk 6 above)
Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule.go]: Updated.

**Review by louyuting — DISMISSED** [review]

File: core/circuitbreaker/rule.go [hunk 8]
```diff
@@ -114,3 +122,16 @@ func (r *Rule) isEqualsTo(newRule *Rule) bool {
 		return false
 	}
 }
+
+func getStatSlidingWindowBucketCount(r *Rule) uint32 {
```
Comment (sanxun0325) [inline core/circuitbreaker/rule.go:126]: Best can complement a for getStatSlidingWindowBucketCount UT

Comment (luckyxiaoqiang) [reply core/circuitbreaker/rule.go]: Already have. I've done some renaming. Please have look again.

**Review by louyuting — APPROVED** [review]
LGTM

**Pull request merged**